lib/customisation.nix

```diff
@@ -117,8 +117,41 @@ rec {
   callPackageWith = autoArgs: fn: args:
     let
       f = if lib.isFunction fn then fn else import fn;
-      auto = builtins.intersectAttrs (lib.functionArgs f) autoArgs;
-    in makeOverridable f (auto // args);
+      fargs = lib.functionArgs f;
+
+      allArgs = builtins.intersectAttrs fargs autoArgs // args;
+      missingArgs = lib.attrNames
+        # Filter out arguments that have a default value
+        (lib.filterAttrs (name: value: ! value)
+        # Filter out arguments that are passed
+        (removeAttrs fargs (lib.attrNames allArgs)));
+
+      getSuggestions = arg: lib.pipe (autoArgs // args) [
+        lib.attrNames
+        # Only use ones that are at most 2 edits away
+        (lib.filter (lib.strings.levenshteinAtMost 2 arg))
+        # Put strings with shorter distance first
+        (lib.sort (x: y: lib.strings.levenshtein x arg < lib.strings.levenshtein y arg))
+        # Only take the first couple results
+        (lib.take 3)
+        # Quote all entries
+        (map (x: "\"" + x + "\""))
+      ];
+
+      prettySuggestions = suggestions:
+        if suggestions == [] then ""
+        else if lib.length suggestions == 1 then ", did you mean ${lib.elemAt suggestions 0}?"
+        else ", did you mean ${lib.concatStringsSep ", " (lib.init suggestions)} or ${lib.last suggestions}?";
+
+      prettyLocation =
+        if lib.isFunction fn then ""
+        else if lib.sources.pathIsDirectory fn then "in ${toString fn}/default.nix "
+        else "in ${toString fn} ";
+
+      errorForArg = arg: "Function ${prettyLocation}"
+        + "called without required argument \"${arg}\"${prettySuggestions (getSuggestions arg)}";
+      error = lib.concatMapStringsSep "\n" errorForArg missingArgs;
+    in if missingArgs == [] then makeOverridable f allArgs else throw error;
```

Comment:
This looks like it won't work for invocations that only become valid *after* overriding. I hope I'm wrong.
Also this is demanding `missingArgs` on all `callPackage` invocations. What's the performance impact?

Reply:
I guess you mean something like this:
```nix
(lib.callPackageWith {} ({ foo }: foo) {}).override { foo = 10; }
```

However this also throws an error even before this change.

Yeah the performance needs to be investigated, I think it's going to be minor though